### PR TITLE
Fix Arm64 Relocator

### PR DESCRIPTION
**Pull request by akiannillo**
* fix can relocate function and add triggering test for Android 7 Arm 64
* little change to work with android gdb

### CODE REVIEW

File: tests/core/interceptor.c [hunk 1]
```diff
@@ -120,6 +119,19 @@ INTERCEPTOR_TESTCASE (attach_to_special_function)
   g_assert_cmpstr (fixture->result->str, ==, ">|<");
 }
 
+INTERCEPTOR_TESTCASE (attach_to_pthread_key_create)
+{
+  gint (* pthread_key_create_impl) (pthread_key_t *key, void (*destructor)(void*));
```
Comment (oleavr) [inline tests/core/interceptor.c]: - Missing space after each of the three last stars.
- Exceeds 80 cols.

File: tests/core/interceptor.c [hunk 2]
```diff
@@ -120,6 +119,19 @@ INTERCEPTOR_TESTCASE (attach_to_special_function)
...
+  pthread_key_t key;
+
+  pthread_key_create_impl = GSIZE_TO_POINTER(
```
Comment (oleavr) [inline tests/core/interceptor.c]: Missing space.

File: tests/core/interceptor.c [hunk 3]
```diff
@@ -120,6 +119,19 @@ INTERCEPTOR_TESTCASE (attach_to_special_function)
...
+      gum_module_find_export_by_name ("libc.so", "pthread_key_create"));
+
+  interceptor_fixture_attach_listener (fixture, 0, pthread_key_create_impl, '>', '<');
+
+  g_assert_cmpint (pthread_key_create_impl (&key, NULL), ==, 0);
```
Comment (oleavr) [inline tests/core/interceptor.c]: Should avoid leaking the key here.

File: tests/core/interceptor.c [hunk 4]
```diff
@@ -120,6 +119,19 @@ INTERCEPTOR_TESTCASE (attach_to_special_function)
   g_assert_cmpstr (fixture->result->str, ==, ">|<");
 }
 
+INTERCEPTOR_TESTCASE (attach_to_pthread_key_create)
```
Comment (oleavr) [inline tests/core/interceptor.c]: To avoid breaking the Windows build this test should be wrapped in:

```
#ifdef G_OS_UNIX
#endif
```

File: Makefile.am [hunk 5]
```diff
@@ -73,12 +73,11 @@ check: push
...
 check-gdb: $(top_builddir)/tests/android-gdb.setup
-	adb shell "${RM} -f /data/local/tmp/tests/debug-socket"
 	adb push "${ANDROID_NDK_ROOT}/prebuilt/android-$(android_arch)/gdbserver/gdbserver" "${RPFX}/tests/"
-	adb shell "${RPFX}/tests/gdbserver +${RPFX}/tests/debug-socket ${RPFX}/tests/gum-tests" &
+	adb shell "${RPFX}/tests/gdbserver :${DEBUG_PORT} ${RPFX}/tests/gum-tests" &
 	sleep 1
-	adb forward tcp:${DEBUG_PORT} localfilesystem:${RPFX}/tests/debug-socket
-	${ANDROID_NDK_ROOT}/toolchains/$(android_toolchain)/prebuilt/darwin-x86_64/bin/$(android_toolprefix)-gdb -x "$(top_builddir)/tests/android-gdb.setup"
+	adb forward tcp:${DEBUG_PORT} tcp:${DEBUG_PORT}
+	${ANDROID_GDB} -x "$(top_builddir)/tests/android-gdb.setup"
```
Comment (oleavr) [inline Makefile.am]: This looks like it belongs in another PR.

File: gum/arch-arm64/gumarm64relocator.c [hunk 6]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+    gpointer key, value;
 
     err = cs_open (CS_ARCH_ARM64, CS_MODE_LITTLE_ENDIAN, &capstone);
     g_assert_cmpint (err, == , CS_ERR_OK);
     err = cs_option (capstone, CS_OPT_DETAIL, CS_OPT_ON);
     g_assert_cmpint (err, ==, CS_ERR_OK);
 
-    count = cs_disasm (capstone, rl.input_cur, 1024, rl.input_pc, 0, &insn);
-    g_assert (insn != NULL);
+    insn = cs_malloc(capstone);
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Missing space.

File: gum/arch-arm64/gumarm64relocator.c [hunk 7]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+    targets_to_check = g_hash_table_new (NULL, NULL);
+    checked_targets = g_hash_table_new (NULL, NULL);
 
-    eoi = FALSE;
-    for (i = 0; i != count && !eoi; i++)
+    code_cur = rl.input_cur;
+    address_cur = rl.input_pc;
+    code_size = 1024;
+    gboolean tragic_exit = false;
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Should always use `TRUE` or `FALSE` for consistency.

File: gum/arch-arm64/gumarm64relocator.c [hunk 8]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+    while (true)
     {
-      cs_arm64 * d = &insn[i].detail->arm64;
+      gboolean go_ahead = true;
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Should always use TRUE or FALSE for consistency.

File: gum/arch-arm64/gumarm64relocator.c [hunk 9]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+      g_hash_table_add (checked_targets, code_cur);
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: This feels like it could be more readable if allowed to breathe a bit, by having a blank line on each side.

File: gum/arch-arm64/gumarm64relocator.c [hunk 10]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+          &address_cur, insn))
+      {
+        cs_arm64 * d = &insn->detail->arm64;
+        switch (insn->id)
+        {
+          case ARM64_INS_B:
+          {
+            cs_arm64_op * op = &d->operands[0];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Missing space.

File: gum/arch-arm64/gumarm64relocator.c [hunk 11]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+            go_ahead = d->cc != ARM64_CC_INVALID && d->cc != ARM64_CC_AL &&
+                d->cc != ARM64_CC_NV;
+            break;
+          }
+          case ARM64_INS_CBZ:
+          case ARM64_INS_CBNZ:
+          {
+            cs_arm64_op * op = &d->operands[1];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Missing space.

File: gum/arch-arm64/gumarm64relocator.c [hunk 12]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+              g_hash_table_add (targets_to_check, op->imm);
+            break;
+          }
+          case ARM64_INS_TBZ:
+          case ARM64_INS_TBNZ:
+          {
+            cs_arm64_op * op = &d->operands[2];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Missing space.

File: gum/arch-arm64/gumarm64relocator.c [hunk 13]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
     }
 
-    cs_free (insn, count);
+    if (!tragic_exit)
+    {
+      g_hash_table_iter_init (&iter, checked_targets);
+      while (g_hash_table_iter_next (&iter, &key, &key))
+      {
+        gssize offset = (gssize) key - (gssize) GPOINTER_TO_SIZE (address);
+        if (offset > 0 && offset < (gssize) n)
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Can't also `offset == 0` if a branch goes to the very first instruction?

File: gum/arch-arm64/gumarm64relocator.c [hunk 14]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+    gboolean tragic_exit = false;
+    while (true)
     {
-      cs_arm64 * d = &insn[i].detail->arm64;
+      gboolean go_ahead = true;
+      g_hash_table_add (checked_targets, code_cur);
+      while (go_ahead && cs_disasm_iter (capstone, &code_cur, &code_size,
+          &address_cur, insn))
+      {
+        cs_arm64 * d = &insn->detail->arm64;
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Could use a blank line here here.

File: gum/arch-arm64/gumarm64relocator.c [hunk 15]
```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+          }
+          case ARM64_INS_BR:
+          {
+            go_ahead = false;
+            tragic_exit = true;
+            /*
+             * tragic exit: we stop and reduce to the safest case because
+             * we can not know where the br can jump. Optimization is possible,
+             * by guessing the register value from previouos instruction.
+             */
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Why is this a tragic exit? There may still be other branches that we encountered previously, right?

Comment (akiannillo) [inline gum/arch-arm64/gumarm64relocator.c]: What I meant it's that once we encountered a BR, it's useless to go ahead or whatever we checked until now. A BR can jump to any position, thus we break the algorithm to return 4.

Comment (akiannillo) [inline gum/arch-arm64/gumarm64relocator.c]: maybe i've be a bit dramatic. Fast exit may sound better

File: gum/arch-arm64/gumarm64relocator.c (hunk 13 above)
Comment (akiannillo) [inline gum/arch-arm64/gumarm64relocator.c]: This came from the old algorithm. My understanding is that we don't care if an instruction jump to the very first instruction, because however we overwrite it, that instruction will jump to the beginning of the trampoline and the behaviour is preserved. Am I wrong?

Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Ahh, yes you're perfectly right, sorry, I had forgotten about that :)

File: gum/arch-arm64/gumarm64relocator.c [hunk 16]
```diff
@@ -346,58 +346,134 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+    targets_to_check = g_hash_table_new (NULL, NULL);
+    checked_targets = g_hash_table_new (NULL, NULL);
 
-    eoi = FALSE;
-    for (i = 0; i != count && !eoi; i++)
+    code_cur = rl.input_cur;
+    address_cur = rl.input_pc;
+    code_size = 1024;
+    gboolean fast_exit = FALSE;
+    while (true)
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: I only pointed out one of these boolean issues to avoid flooding the PR with comments. There are also other left-overs. (Just search for `true` and `false`.)

File: gum/arch-arm64/gumarm64relocator.c [hunk 17]
```diff
@@ -346,58 +346,134 @@ gum_arm64_relocator_can_relocate (gpointer address,
...
+    insn = cs_malloc (capstone);
+    targets_to_check = g_hash_table_new (NULL, NULL);
+    checked_targets = g_hash_table_new (NULL, NULL);
 
-    eoi = FALSE;
-    for (i = 0; i != count && !eoi; i++)
+    code_cur = rl.input_cur;
+    address_cur = rl.input_pc;
+    code_size = 1024;
+    gboolean fast_exit = FALSE;
```
Comment (oleavr) [inline gum/arch-arm64/gumarm64relocator.c]: Declaration after statement.

**Comment by oleavr** [comment]
Also there's a lot of compiler warnings introduced by this PR – did you ignore them or are you using a different toolchain than me?

**Comment by oleavr** [comment]
Tweaked version merged in f59b0be7f10369b5f34ce97beda86daa493623f7. Thanks!
